src/AltinnCore/Runtime/RestControllers/ProcessController.cs

```diff
@@ -41,29 +66,445 @@ public class ProcessController : ControllerBase
+
+            instance.Process = new ProcessState
+            {
+                Started = now,
+                StartEvent = validStartElement,
+            };
+            Instance updatedInstance = await instanceService.UpdateInstance(instance, app, org, instanceOwnerId, instanceGuid);
+            List<InstanceEvent> events = new List<InstanceEvent>
+            {
+                GenerateProcessChangeEvent("process:StartEvent", updatedInstance, validStartElement, now)
+            };
+
+            await DispatchEvents(org, app, events);
+
+            return updatedInstance;
+        }
+
+        private async Task<Instance> UpdateProcessStateToNextElement(string org, string app, Instance instance, int instanceOwnerId, Guid instanceGuid, string nextElementId)
+        {
+            List<InstanceEvent> events = ChangeProcessState(instance, nextElementId);
+
+            Instance changedInstance = await instanceService.UpdateInstance(instance, app, org, instanceOwnerId, instanceGuid);
+
+            await DispatchEvents(org, app, events);
+
+            return changedInstance;
+        }
+
+        private async Task DispatchEvents(string org, string app, List<InstanceEvent> events)
+        {
+            foreach (InstanceEvent instanceEvent in events)
+            {
+                await eventService.SaveInstanceEvent(instanceEvent, org, app);
+            }
+        }
+
+        private string CheckOrGetValidNextElement(string currentElementId, string proposedElementId, out ActionResult nextElementError)
+        {
+            nextElementError = null;
+            string nextElement = null;
+
+            List<string> possibleNextElements = ProcessKeeper.NextElements(currentElementId);
+            
+            if (!string.IsNullOrEmpty(proposedElementId) && possibleNextElements.Contains(proposedElementId))
+            {
+                nextElement = proposedElementId;
+            }
+            else if (possibleNextElements.Count == 1)
+            {
+                nextElement = possibleNextElements.First();
+            }
+            else if (!string.IsNullOrEmpty(proposedElementId))
+            {
+                nextElementError = Conflict($"Process element id '{proposedElementId}' is not found in app's process model (bpmn)");
+            }
+            else if (possibleNextElements.Count > 1)
+            {
+                nextElementError = Conflict($"There are more than one outgoing sequence flow, please select one '{possibleNextElements}'");
+            }
+
+            return nextElement;
+        }
+
+        /// <summary>
+        ///  Todo: Fix validation code of the instance's currentTask.
+        /// </summary>        
+        /// <returns>try if validation is OK, false otherwise</returns>
+        private bool InstanceIsValid(Instance instance)
+        {
+            return true;            
+        }
+
+        private List<InstanceEvent> ChangeProcessState(Instance instance, string nextElementId)
```

Comment:
This function both moves the instance forward in the process and generates events. Could we either make sure the function only has one responsibility or give it a more suitable name?